File: examples/solvablepolynomial.rb
#
# jruby examples for jas.
# $Id$
#

require "examples/jas"

# WA_32 solvable polynomial example

p = PolyRing.new(QQ(),"a,b,e1,e2,e3");
#is automatic: [one,a,b,e1,e2,e3] = p.gens();

relations = [e3, e1, e1*e3 - e1,
             e3, e2, e2*e3 - e2];

puts "relations: =" + relations.join(", ") { |r| r.to_s };
puts;


rp = SolvPolyRing.new(QQ(), "a,b,e1,e2,e3", PolyRing.lex, relations);
puts "SolvPolyRing: " + str(rp);
puts;

puts "gens =" + rp.gens().join(", ") { |r| r.to_s };
one,a,b,e1,e2,e3 = rp.gens();

f1 = e1 * e3**3 + e2**10 - a;
f2 = e1**3 * e2**2 + e3;
f3 = e3**3 + e3**2 - b;

#puts "f1 = ", f1;
#puts "f2 = ", f2;
#puts "f3 = ", f3;

F = [ f1, f2, f3 ];
puts "F =" + F.join(", ") { |r| r.to_s };
puts

I = rp.ideal( "", F );
puts "SolvableIdeal: " + str(I);
puts;

rgl = I.leftGB();
puts "seq left GB:" + str(rgl);
puts "isLeftGB: " + str(rgl.isLeftGB());
puts;

rgt = I.twosidedGB();
puts "seq twosided GB:" + str(rgt);
puts "isTwosidedGB: " + str(rgt.isTwosidedGB());
puts;

rgr = I.rightGB();
puts "seq right GB:" + str(rgr);
puts "isRightGB: " + str(rgr.isRightGB());
puts;

#exit(0);


rs = """
# solvable polynomials, Weyl algebra A_3,2:
Rat(a,b,e1,e2,e3) L
#Quat(a,b,e1,e2,e3) G|3|
RelationTable
(
 ( e3 ), ( e1 ), ( e1 e3 - e1 ),
 ( e3 ), ( e2 ), ( e2 e3 - e2 )
)
""";

r = SolvableRing.new( rs );
puts "SolvableRing: " + str(r);
puts;

#one,a,b,e1,e2,e3 = r.gens();

f1 = e1 * e3**3 + e2**10 - a;
f2 = e1**3 * e2**2 + e3;
f3 = e3**3 + e3**2 - b;

puts "f1: " + str(f1);
puts "f2: " + str(f2);
puts "f3: " + str(f3);
puts;

ff = r.ideal( "", [f1,f2,f3] );
puts "SolvableIdeal: " + str(ff);
puts;

#exit();

rg = ff.leftGB();
puts "seq left GB: " + str(rg);
puts "isLeftGB: " + str(rgt.isLeftGB());
puts;
puts "rgl == rg: " + str(rgl.list == rg.list);
puts;

#exit();

rg = ff.twosidedGB();
puts "seq twosided GB: " + str(rg);
puts "isTwosidedGB: " + str(rgt.isTwosidedGB());
puts;
puts "rgt == rg: " + str(rgt.list == rg.list);


rg = ff.rightGB();
puts "seq right GB: " + str(rg);
puts "isRightGB: " + str(rgt.isRightGB());
puts;
puts "rgr == rg: " + str(rgr.list == rg.list);

#startLog();
#terminate();

File: examples/solvablepolynomial.py
#
# jython examples for jas.
# $Id$
#

from java.lang import System
from java.lang import Integer

from jas import SolvableRing, SolvPolyRing, PolyRing, SolvableIdeal
from jas import QQ, Quat

# WA_32 solvable polynomial example

p = PolyRing(QQ(),"a,b,e1,e2,e3");
#is automatic: [one,a,b,e1,e2,e3] = p.gens();

relations = [e3, e1, e1*e3 - e1,
             e3, e2, e2*e3 - e2];

print "relations: =", [ str(f) for f in relations ];
print;

#rp = SolvPolyRing(QQ(), "a,b,e1,e2,e3", rel=relations);
rp = SolvPolyRing(QQ(), "a,b,e1,e2,e3", PolyRing.lex, relations);
print "SolvPolyRing: " + str(rp);
print;

print "gens =", [ str(f) for f in rp.gens() ];
#[one,a,b,e1,e2,e3] = rp.gens();
#[one,I,J,K,a,b,e1,e2,e3] = rp.gens();

f1 = e1 * e3**3 + e2**10 - a;
f2 = e1**3 * e2**2 + e3;
f3 = e3**3 + e3**2 - b;

f4 = ( e3**2 * e2**3 + e1 )**3;

#print "f1 = ", f1;
#print "f2 = ", f2;
#print "f3 = ", f3;
#print "f4 = ", f4;

F = [ f1, f2, f3 ];
print "F =", [ str(f) for f in F ];
print

I = rp.ideal( list=F );
print "SolvableIdeal: " + str(I);
print;

rgl = I.leftGB();
print "seq left GB:" + str(rgl);
print "isLeftGB: " + str(rgl.isLeftGB());
print;

rgt = I.twosidedGB();
print "seq twosided GB:" + str(rgt);
print "isTwosidedGB: " + str(rgt.isTwosidedGB());
print;

rgr = I.rightGB();
print "seq right GB:" + str(rgr);
print "isRightGB: " + str(rgr.isRightGB());
print;


#exit(0);


rs = """
# solvable polynomials, Weyl algebra A_3,2:
Rat(a,b,e1,e2,e3) L
#Quat(a,b,e1,e2,e3) G|3|
RelationTable
(
 ( e3 ), ( e1 ), ( e1 e3 - e1 ),
 ( e3 ), ( e2 ), ( e2 e3 - e2 )
)
""";

r = SolvableRing( rs );
print "SolvableRing: " + str(r);
print;

print "gens =", [ str(f) for f in r.gens() ];
#[one,a,b,e1,e2,e3] = r.gens();
#[one,I,J,K,a,b,e1,e2,e3] = r.gens();

fs1 = e1 * e3**3 + e2**10 - a;
fs2 = e1**3 * e2**2 + e3;
fs3 = e3**3 + e3**2 - b;

fs4 = ( e3**2 * e2**3 + e1 )**3;

#print "fs1 = ", fs1;
#print "fs2 = ", fs2;
#print "fs3 = ", fs3;
print "fs4 = ", fs4;

Fs = [ fs1, fs2, fs3 ];
print "Fs =", [ str(f) for f in Fs ];
print

Is = r.ideal( list=Fs );
print "SolvableIdeal: " + str(Is);
print;

rgsl = I.leftGB();
print "seq left GB:" + str(rgsl);
print "isLeftGB: " + str(rgsl.isLeftGB());
print;

rgst = I.twosidedGB();
print "seq twosided GB:" + str(rgst);
print "isTwosidedGB: " + str(rgst.isTwosidedGB());
print;

rgsr = I.rightGB();
print "seq right GB:" + str(rgsr);
print "isRightGB: " + str(rgsr.isRightGB());
print;


print "rgl == rgsl: " + str(rgl == rgsl);
print "rgr == rgsr: " + str(rgr == rgsr);
print "rgt == rgst: " + str(rgt == rgst);
